libursa/src/signatures/bls.rs

```diff
@@ -0,0 +1,706 @@
+/// Implements
+/// https://eprint.iacr.org/2018/483 and
+/// https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html
+use amcl_wrapper::{
+    constants::{GroupG1_SIZE, MODBYTES},
+    extension_field_gt::GT,
+    field_elem::FieldElement,
+    group_elem::GroupElement,
+    group_elem_g1::G1,
+    group_elem_g2::G2,
+    types_g2::GroupG2_SIZE,
+};
+use CryptoError;
+
+pub const PRIVATE_KEY_SIZE: usize = MODBYTES;
+/// This is a simple alias so the consumer can just use PrivateKey::random() to generate a new one
+/// instead of wrapping it as a private field
+pub type PrivateKey = FieldElement;
+
+macro_rules! public_key_impl {
+    () => {
+        #[derive(Debug, Clone, Serialize, Deserialize)]
+        pub struct PublicKey(Generator);
+
+        impl PublicKey {
+            pub fn new(sk: &PrivateKey, g: &Generator) -> Self {
+                PublicKey(g * sk)
+            }
+
+            // Create an combined public key without rogue key mitigation
+            pub fn combine(&mut self, pks: &[PublicKey]) {
+                for pk in pks {
+                    self.0 += &pk.0;
+                }
+            }
+
+            pub fn to_bytes(&self) -> Vec<u8> {
+                self.0.to_bytes()
+            }
+
+            pub fn from_bytes(bytes: &[u8]) -> Result<Self, CryptoError> {
+                Ok(PublicKey(Generator::from_bytes(bytes).map_err(|e| {
+                    CryptoError::ParseError(format!("{:?}", e))
+                })?))
+            }
+        }
+    };
+}
+
+macro_rules! aggregate_public_key_impl {
+    () => {
+        /// Represents an aggregated BLS public key that mitigates the rogue key attack
+        /// for verifying aggregated signatures.
+        #[derive(Debug, Clone, Serialize, Deserialize)]
+        pub struct AggregatedPublicKey(Generator);
+
+        impl From<&[PublicKey]> for AggregatedPublicKey {
+            fn from(keys: &[PublicKey]) -> Self {
+                // To combat the rogue key attack,
+                // compute (t_1,…,t_n)←H1(pk_1,…,pk_n) ∈ R_n
+                // output the aggregated public key
+                // as described in section 3.1 from https://eprint.iacr.org/2018/483
+                let mut bytes = Vec::new();
+                for k in keys {
+                    bytes.extend_from_slice(k.to_bytes().as_slice());
+                }
+                AggregatedPublicKey(keys.iter().fold(Generator::identity(), |apk, k| {
+                    // The position of the ith public key in the byte array
+                    // of the hash doesn't matter as much as its included twice.
+                    // For convenience, its appended to the end
+                    let mut h = bytes.clone();
+                    h.extend_from_slice(k.0.to_bytes().as_slice());
+                    apk + &k.0 * &FieldElement::from_msg_hash(h.as_slice())
```

Comment:
Domain separation?

Reply:
I don’t think it’s necessary just like for simple signatures.

Reply:
For what kind of signatures do you think it's necessary? FYI, the paper says that you should use different hash functions for hashing verkeys and messages (H1 and H0 respectively)

Reply:
It's up to the caller to determine the domains to use. The caller can prepend or append the domain separation to the message that's being signed. That's no different than Ed25519, Secp256k1, or RSA.

Reply:
The domains are determined, one is the message's and the other is verkey's. Why does the caller cares whether you simulate different random oracles using different hash functions or using different prefix with the same hash function?